Fixes #37987 - Filters remove override, unlimted, and taxonomies

MariaAga opened the pull request:
To test: 
1. Create a role
2. Assign an organization and a location to the role
3. Create a filter under the role with host permissions 
before the pr: no taxonomy_search in the DB, and the filter is marked as unlimited 

Added a migration to regenerate the taxonomy_search field of all filters which have override set to false to make them honor the organizations and locations set on the role. 

Added a ui block on the unlimited checkbox if there are taxonomies in place.

Fixed "build_taxonomy_search" as it relayed on `locations`, but on creation there are only `location_ids`. Before this pr the workaround would be to edit and submit the filter (with no changes) since on edit `locations` are already populated

adamruzicka commented on app/models/filter.rb:
```diff
@@ -165,7 +165,22 @@ def enforce_inherited_taxonomies
     self.organization_ids = role.organization_ids if resource_taxable_by_organization?
     self.location_ids = role.location_ids if resource_taxable_by_location?
-    build_taxonomy_search
+    build_taxonomy_search_from_ids
```
Would keeping this and changing L198 in the new diff to use map instead of pluck help?

MariaAga replied:
Changing just `build_taxonomy_search` seem to work as well for create and edit filter

ofedoren commented on db/migrate/20241105143353_regenerate_taxonomy_search_for_filters_with_override_false.rb:
```diff
@@ -0,0 +1,10 @@
+    filters = Filter.where(role_id: Role.where(origin: nil).or(Role.where(builtin: 2))).where(override: false).where(taxonomy_search: nil)
+
+    filters.each do |filter|
+      filter.build_taxonomy_search_from_ids
```
```suggestion
      filter.build_taxonomy_search
```

ofedoren commented on test/models/filter_test.rb:
```diff
@@ -245,8 +245,17 @@ class FilterTest < ActiveSupport::TestCase
+    location_test = FactoryBot.create(:location)
+    f.role = FactoryBot.create(:role, :organizations => [organization_test], :locations => [location_test])
+    f.save
+    assert_equal f.taxonomy_search(), "(organization_id ^ (#{organization_test.id})) and (location_id ^ (#{location_test.id}))"
```
```suggestion
    assert_equal f.taxonomy_search, "(organization_id ^ (#{organization_test.id})) and (location_id ^ (#{location_test.id}))"
```

ofedoren commented on webpack/assets/javascripts/react_app/routes/FiltersForm/NewFiltersFormPage.js:
```diff
@@ -29,7 +29,7 @@ const NewFiltersFormPage = ({ location: { search }, history }) => {
           roleName={roleName || ''}
           roleId={urlRoleId}
-          data={{}}
+          data={{ locations, organizations }}
```
Does
```js
data.organizations?.map(o => o.id) || []
```
mean it would always be `[]` since taxonomies weren't passed within `data` before your patch?..

MariaAga replied:
This is the new page, in the edit page we have `data={response}`

MariaAga commented:
thanks @ofedoren, updated

adamruzicka commented on app/models/filter.rb:
```diff
@@ -168,24 +168,24 @@ def inherit_taxonomies!
+  def build_taxonomy_search_string_from_ids(name)
     return '' unless send("resource_taxable_by_#{name}?")
-    relation = send(name.pluralize).pluck(:id)
+    relation = send("#{name}_ids")
```
_nitpick:_ is this rename necessary?

MariaAga replied:
Probably leftover from when that pr had more methods, removed the remane

adamruzicka commented on app/models/filter.rb:
```diff
@@ -168,24 +168,24 @@ def inherit_taxonomies!
     parenthesize("#{name}_id ^ (#{relation.join(',')})")
   end
 
+  private
+
```
Was this moved here so that you can call `build_taxonomy_search` in the migration?

MariaAga replied:
Probably

adamruzicka replied:
That doesn't seem to be necessary anymore, could we move it back?

MariaAga commented:
Draft until I finish  "remove taxonimoes override from filter"

aneta-petrova commented on webpack/assets/javascripts/react_app/routes/FiltersForm/FiltersForm.js:
```diff
@@ -117,7 +130,29 @@ export const FiltersForm = ({ roleName, roleId, isNew, data, history }) => {
+
+      <FormGroup
+        label={__('Locations')}
+        helperText={__('Taxonomies are inherited from the role')}
```
Instead of "Taxonomies are inherited...", how about adding a line like that after Organizations and Locations each? So: "Organizations are inherited..." and "Locations are inherited ...". It's two lines instead of one but I'm concerned about the term taxonomies: will everyone know that it refers to orgs and locs?

Also, looking at the web UI, due to the layout of the page (indentation) and the formatting, it might look like "Taxonomies are inherited..." relates only to Locations.

MariaAga replied:
Yeah, thats a better way, I'll change it

ofedoren commented on app/controllers/concerns/foreman/controller/parameters/filter.rb:
```diff
@@ -8,9 +8,7 @@ def filter_params_filter
           :unlimited,
-          :override,
           :permissions => [], :permission_ids => [], :permission_names => []
         add_taxonomix_params_filter(filter)
```
```suggestion
```
along with
```ruby
  include Foreman::Controller::Parameters::Taxonomix
```

ofedoren commented on app/models/filter.rb:
```diff
@@ -1,7 +1,6 @@
 class Filter < ApplicationRecord
   audited :associated_with => :role
 
-  include Taxonomix
```
~IIUC, we don't make filters non-taxonomable, but rather force them to be in line with the role they're assigned to disabling user override possibility.~

~If so, this should stay since this concern makes filter model taxonomable.~

I think by disabling possibility to override taxonomies on filter level, there is no longer need to have taxonomy association, `taxonomy_search` can be created on the fly during `before_save`.

MariaAga replied:
isnt that what `before_save :enforce_inherited_taxonomies,` do?

ofedoren reviewed (commented):
There are some comments, you can ignore until the final review (pasting them just so I don't lose them)

aneta-petrova commented on webpack/assets/javascripts/react_app/routes/FiltersForm/FiltersForm.js:
```diff
@@ -117,54 +117,44 @@ export const FiltersForm = ({ roleName, roleId, isNew, data, history }) => {
+                ouiaId="taxonomy-search-alert"
+                variant="info"
+                title={__(
+                  "The filter is scoped to the selected organizations and locations, therefore can't be unlimited"
```
Not sure if helpful or just nitpicking, but I actually had trouble parsing what the alert meant. Do you think this would be better?

```suggestion
                  "You cannot use unlimited filtering because the filter is scoped to the organizations and locations inherited from the role."
```

I propose s/selected/inherited because 1. I think you don't really "select" orgs/locs anymore and 2. "inherited" is used in the descriptions on lines 122 and 135, I think consistency would help users connect the dots more quickly.

aneta-petrova replied:
Would it be better to move the info alert below the (greyed-out) unlimited checkbox?

MariaAga replied:
Yeah, for consistency should we change the warning to be a field small warning or an alert? 
The 2 options shown here:
![image](https://github.com/user-attachments/assets/e7ed1480-7240-4a5c-9463-9aa69738427e)

MariaAga replied:
@aneta-petrova ?

aneta-petrova replied:
Sorry, I have a hard time deciding which of the two options would be better. What do you think?

I guess what I feel most strongly about here is the wording -- as suggested above.

ofedoren commented on app/models/role.rb:
```diff
@@ -216,10 +216,6 @@ def remove_permissions!(*args)
   end
 
-  def disable_filters_overriding
-    filters.where(:override => true).map { |filter| filter.disable_overriding! }
```
Could we also remove `.where(:override => false)` in `sync_inheriting_filters` below in this file?

ofedoren commented on db/migrate/20250109165511_remove_taxonomy_and_override_from_filter.rb:
```diff
@@ -0,0 +1,23 @@
+class RemoveTaxonomyAndOverrideFromFilter < ActiveRecord::Migration[7.0]
+  def up
+    # remove_column :filters, :taxonomy_search
```
Do we want to remove the column even though we use it?

MariaAga replied:
Probably not? we want to keep using it right?

ofedoren commented on db/migrate/20250109165511_remove_taxonomy_and_override_from_filter.rb:
```diff
@@ -0,0 +1,23 @@
+  def up
+    # remove_column :filters, :taxonomy_search
+    # remove_column :filters, :override
+    # filters = Filter.where(role_id: Role.where(origin: nil).or(Role.where(builtin: 2)))
```
Out of curiosity: if we drop overriding anyway and force all the filters use inherited taxonomies, shouldn't we update all the filters?

MariaAga replied:
I think this selects only the editable filters

ofedoren replied:
> I think this selects only the editable filters

Aren't they all editable?..

Also, since filters were taxonomable and we're removing this, we should probably also clean related leftovers, such as `TaxableTaxonomy.where(taxable_type: 'Filter')`.

MariaAga replied:
Not the locked ones 
```rb
  def locked?
    return false if modify_locked || self.class.modify_locked
    return false unless respond_to? :origin
    origin.present? && builtin != BUILTIN_DEFAULT_ROLE
  end

```

ofedoren replied:
But this is for `role`, not `filter`?

MariaAga replied:
yes but filters are a part of one role, and if its locked it cant be edited, not the role and not its filters, so it cant have override: true
Or do we just want to run it on the default roles anyway?

ofedoren replied:
My apologies, you're right. It a role is locked, the filters can't be overriden and we also don't allow to lock roles.

I guess we could then search for overriden filters before we remove the column to update their taxonomy search to be based on the assigned role. And after that we could remove the column and the related taxonomy records.

MariaAga replied:
Since this bug is also for "Newly created filters do not inherit organizations and locations associated with the role"
is the search ok or should we change the migration to something else? @ofedoren

ofedoren replied:
Hmm.. Firstly I thought to go over overridden filters only, but your current search gathers all the filters which don't have a locked role attached (thus with overridden filters as well) and this way we force new behavior on all user facing filters, which is safer I guess.

I'd say the migration is correct, but I'd ask for another opinion just in case, @adamruzicka.

adamruzicka replied:
It seems to be correct. I even thought about going overboard with this and rebuilding all the filters, with the assumption that for most of the filters it will be a noop. But as you said, the current search should be fine.

ofedoren reviewed (commented):
Thanks, @MariaAga, LGTM, just few cents inline (consider previous comments as well). Small rubocop issue as well https://github.com/theforeman/foreman/actions/runs/12709621398/job/35429016546?pr=10370

UPD: Reminder for me to remove related stuff from hammer: https://github.com/theforeman/hammer-cli-foreman/blob/master/lib/hammer_cli_foreman/filter.rb#L13

ofedoren reviewed (commented):
@MariaAga, can this be undrafted after the migration is uncommented or you're planning to do more changes?

Also, while checking something, I've found we have another, but rather small, PF5 related issue here: https://github.com/theforeman/foreman/blob/develop/webpack/assets/javascripts/react_app/components/common/Pf4DualList/DualListWithIds.js#L42

it should be
```js
const onListChange = (_e, newAvailableOptions, newChosenOptions) => {
```
now

UPD: just noticed, this needs a rebase.

adamruzicka commented on db/migrate/20250109165511_remove_taxonomy_and_override_from_filter.rb:
```diff
@@ -0,0 +1,22 @@
+    # filters = Filter.where(role_id: Role.where(origin: nil).or(Role.where(builtin: 2)))
+
+    # filters.each do |filter|
+    #   filter.enforce_inherited_taxonomies
+    #   filter.update_column(:taxonomy_search, filter.taxonomy_search)
```
After `filter.enforce_inherited_taxonomies`, the `taxonomy_search` field already contains the right value. Why go with `update_column` instead of just saving the whole record?

MariaAga replied:
changed to `filter.save`

ofedoren replied:
@adamruzicka, wouldn't `save` cause all the validations and callbacks to be run, thus making `enforce_inherited_taxonomies` run twice whilst `update_column` will just put a new value into DB?

adamruzicka replied:
Oh, yes, nevermind

MariaAga commented:
Thanks for the reviews, I rebased and addressed the comments

> can this be undrafted after the migration is uncommented or you're planning to do more changes?

Yes, migration is just scary

ofedoren commented on db/migrate/20241105143353_regenerate_taxonomy_search_for_filters_with_override_false.rb:
```diff
@@ -0,0 +1,10 @@
+class RegenerateTaxonomySearchForFiltersWithOverrideFalse < ActiveRecord::Migration[6.1]
+  def change
+    filters = Filter.where(role_id: Role.where(origin: nil).or(Role.where(builtin: 2))).where(override: false).where(taxonomy_search: nil)
```
```ruby
    where(taxonomy_search: nil)
```
Was it in the previous iteration of commit? I thought we wanted to explicitly override any overridden searches, this will only add instead thus leaving overridden searches, which seems wrong...

ofedoren reviewed (commented):
Thanks, @MariaAga, but this seems to have less changes then before, for example all changes from e.g. https://github.com/theforeman/foreman/pull/10370#pullrequestreview-2546689522 are gone. Is that intentional? At very least leaving `enforce_override_flag` seems against the idea of the PR.

MariaAga commented:
Revival! 
Seems that one commit wasnt pushed? I combined them,rebased, updated to pf5 
I also removed unlimited from the form as it seems like a confusing field that just says "check this if you want the search bar to be enabled"?
I noticed that roles are not filtered by taxonomy before this pr anyway so I'm not sure what it means for all of this pr 
Also moved the Taxonomies to common since we'll need then for page refactors

MariaAga commented on webpack/assets/javascripts/react_app/routes/FiltersForm/FiltersForm.js:
```diff
@@ -118,93 +101,40 @@ export const FiltersForm = ({ roleName, roleId, isNew, data, history }) => {
+          aria-label="disabled organization field"
+          isDisabled
+          type="text"
+          value={chosenOrgs.join(' ,')}
```
no, thats for readability 
<img width="236" height="193" alt="image" src="https://github.com/user-attachments/assets/a370c064-11d7-4240-8a92-e825732120ca" />

ofedoren commented on db/migrate/20250109165511_remove_taxonomy_and_override_from_filter.rb:
```diff
@@ -0,0 +1,22 @@
+  end
+
+  def down
+    # add_column :filters, :taxonomy_search, :text
```
This is not needed, we don't remove that column anyway.

ofedoren commented on db/migrate/20250109165511_remove_taxonomy_and_override_from_filter.rb:
```diff
@@ -0,0 +1,22 @@
+    # filters = Filter.where(role_id: Role.where(origin: nil).or(Role.where(builtin: 2))).where(override: false).where(taxonomy_search: nil)
+
+    # filters.each do |filter|
+    #   filter.build_taxonomy_search
```
Maybe it's better to use the same logic from `up`?
```suggestion
    #   filter.enforce_inherited_taxonomies
```

ofedoren reviewed (commented):
Thanks, @MariaAga ! I don't have anything for backend side, just two cents for migration, which should be uncommented :)

It probably also needs a release note since it's quite a breaking change.

MariaAga commented:
@ofedoren is it ok that the pr also removes the override checkbox from the ui?

ofedoren commented:
> @ofedoren is it ok that the pr also removes the override checkbox from the ui?

Is that... unexpected? I mean, we forbid filters to be overridden, so the checkbox would be misleading, wouldn't it? And essentially no-op... So I'd say it's a must?..

MariaAga commented:
>Is that... unexpected? I mean, we forbid filters to be overridden, so the checkbox would be misleading, wouldn't it? And essentially no-op... So I'd say it's a must?..

Not sure what you mean?
Right now it works like "Im a user, I want to add search, I first need to enable the search box by checking this box, and then I can type in my search"
So we can just skip the checkbox and define "override" by if there is a search present

ofedoren commented:
> Not sure what you mean? Right now it works like "Im a user, I want to add search, I first need to enable the search box by checking this box, and then I can type in my search" So we can just skip the checkbox and define "override" by if there is a search present

Wait, do you mean `unlimited` checkbox? I've just noticed in UI code you've removed that... why? We just removing `override`, which is related to overriding taxonomies scope. Unlimited is something not related, thus should be left.

MariaAga commented:
I'm not sure about keeping the unlimited field in the ui form/ at all? since now no filters will be unlimited since they will always have a `taxonomy_search`

ofedoren commented:
> I'm not sure about keeping the unlimited field in the ui form/ at all? since now no filters will be unlimited since they will always have a `taxonomy_search`

But taxonomy search is only for scoping by taxonomies. Search on the other hand if for limiting scope for other resources, which is dependent on `unlimited` check.

MariaAga commented:
So should we change unlimited to ignore taxonomy_search? 
https://github.com/theforeman/foreman/blob/0d9d5fddf0f326b7d882f90d84e8061445fe0b23/app/models/filter.rb#L92-L94
`scope :unlimited, -> { where(:search => nil, :taxonomy_search => nil) }`

ofedoren commented:
Why? Do roles always have taxonomies attached?

MariaAga commented:
when a user creates a role they choose a taxonomy, and in this pr we do
```
   filters.each do |filter|
       filter.enforce_inherited_taxonomies
       filter.update_column(:taxonomy_search, filter.taxonomy_search)
```

ofedoren commented:
> So should we change unlimited to ignore taxonomy_search?

Per `unlimited` description: 
> 'If the unlimited field is enabled, the filter applies to all resources of the selected type. If the unlimited field is disabled, you can specify further filtering using Foreman search syntax in the search field. If the role is associated with organizations or locations, the filters are not considered unlimited as they are scoped accordingly.'

So, I guess, no?

Even though we force filters to have roles' taxonomies, it doesn't mean we force roles to have taxonomies.

MariaAga commented:
What is the purpose of a role without any taxonomies?  it doesnt mean "all taxonomies", so what is its use?

MariaAga committed:
Fixes #37987 - Filters remove override, unlimted, and taxonomies

MariaAga commented:
As we agreed, I removed the unlimited field as well

MariaAga commented:
Test failures are unrelated

MariaAga commented on webpack/assets/javascripts/react_app/routes/FiltersForm/FiltersForm.js:
```diff
@@ -4,54 +4,40 @@ import { useDispatch } from 'react-redux';
-    show_organizations: showOrgs = false,
-    show_locations: showLocations = false,
-  } = type;
+  const chosenOrgs = data?.role?.organizations?.map(o => o.name) || [];
+  const chosenLocations = data?.role?.locations?.map(l => l.name) || [];
```
This is readonly information for the user to see which orgs are applied

MariaAga commented on app/models/filter.rb:
note: make sure this pr doesnt clash with the changes https://github.com/theforeman/foreman/pull/10701

adamruzicka replied:
I'd rather get this one in first and then adjust 10701.

adamruzicka commented:
Could you please rebase to make the tests turn green?

adamruzicka commented on db/migrate/20250109165511_remove_taxonomy_and_override_from_filter.rb:
```diff
@@ -0,0 +1,19 @@
+class RemoveTaxonomyAndOverrideFromFilter < ActiveRecord::Migration[7.0]
+  def up
+    remove_column :filters, :override
+    filters = Filter.where(role_id: Role.where(origin: nil).or(Role.where(builtin: 2)))
```
Since we're dropping Filter - Organization/Location association, we should also prune the existing records from taxable_taxonomies table.

MariaAga replied:
Isnt it done here? https://github.com/theforeman/foreman/blob/8c2f07fc3a7b949db6aad8e925aeda04bbab2da7/db/migrate/20210609093404_drop_override_taxonomies_from_filter.rb#L6

MariaAga replied:
Should I do something with 
```  rb
def self.allows_taxonomy_filtering?(_taxonomy)
    false
  end
``` 
?

adamruzicka replied:
> Isnt it done here?

That migration only removes the filter-taxonomy link for filters related to 4 specific permissions. In here, we need to do it for all the filters, no matter which resource and permissions they're linked to.

> Should I do something with `def self.allows_taxonomy_filtering?(_taxonomy)`

No, that should stay I think.

MariaAga replied:
will it look like: `TaxableTaxonomy.where(taxable:Filter.all).delete_all` ?

adamruzicka replied:
`TaxableTaxonomy.where(taxable_type: 'Filter').delete_all` should do

adamruzicka commented on app/models/filter.rb:
```diff
@@ -153,42 +127,32 @@ def expire_topbar_cache
-    self.location_ids = role.location_ids if resource_taxable_by_location?
+    @organization_ids = role.organization_ids if resource_taxable_by_organization?
+    @location_ids = role.location_ids if resource_taxable_by_location?
     build_taxonomy_search
```
Rather than setting up instance variables to be used somewhere later, I'd prefer to pass those along explicitly to make it easier to follow.

```suggestion
    organization_ids = role.organization_ids if resource_taxable_by_organization?
    location_ids = role.location_ids if resource_taxable_by_location?
    build_taxonomy_search(organization_ids, location_ids)
```

Essentially what copilot said a bit lower https://github.com/theforeman/foreman/pull/10370/files#r2269497669

adamruzicka commented on app/models/filter.rb:
```diff
@@ -153,42 +127,32 @@ def expire_topbar_cache
 
-  def build_taxonomy_search_string(name)
+  def build_taxonomy_search_string_from_ids(name)
     return '' unless send("resource_taxable_by_#{name}?")
```
This is already checked at L131-132.
```suggestion
```

adamruzicka reviewed (changes requested):
The things to change are fairly minor, works well apart from those.

ofedoren commented on app/models/filter.rb:
```diff
@@ -153,42 +127,31 @@ def expire_topbar_cache
-    return '' unless send("resource_taxable_by_#{name}?")
-    relation = send(name.pluralize).pluck(:id)
+  def build_taxonomy_search_string_from_ids(name, ids)
+    relation = ids
```
This is not needed, `ids` can be just used down below instead.

ofedoren reviewed (dismissed):
Thanks, @MariaAga and @adamruzicka, I'm afraid I have nothing to add, but one nitpick that can be ignored :)

UPD: I've created a ticket for hammer part: https://projects.theforeman.org/issues/38805. I'll probably pick this myself.

UPD2: Ah... Since we're touching API (and hammer), I'm afraid we might need to touch FAM as well :/ Not that sure though.

adamruzicka commented:
> UPD2: Ah... Since we're touching API (and hammer), I'm afraid we might need to touch FAM as well :/ Not that sure though.

I already looked. Unless I'm misreading it, the functionality that is being removed was never implemented there so we shouldn't need to do anything.

adamruzicka commented:
Thank you @MariaAga & @ofedoren !
